R/dist.R

```diff
@@ -821,57 +867,242 @@ tune_inv_gamma <- function(lower = 2, upper = 21) {
-                      dist = c("lognormal", "gamma"), max = NULL,
-                      fixed = FALSE) {
-  dist <- match.arg(dist)
+                      dist = c("lognormal", "gamma"), max,
+                      pmf = numeric(0), fixed = FALSE) {
+  ## check if parametric or nonparametric
+  if (length(pmf) > 0 &&
+        !all(missing(mean), missing(sd), missing(mean_sd),
+          missing(sd_sd), missing(dist), missing(max))) {
+    stop("Distributional parameters or a pmf can be specified, but not both.")
+  }
 
-  if (missing(mean)) {
+  if (fixed) {
+    mean_sd <- 0
+    sd_sd <- 0
+  }
+  fixed <- mean_sd == 0 && mean_sd == 0
+
+  ## check parametric parameters make sense
+  if (!missing(mean)) {
+    if (sd == 0 && sd_sd == 0) { ## integer fixed
+      if (mean %% 1 != 0) {
+        stop(
+          "When a distribution is set to a constant ",
+          "(sd == 0 and sd_sd == 0) then the mean parameter ",
+          "must be an integer."
+        )
+      }
+      max <- mean
+      if (mean_sd > 0) {
+        stop(
+          "When a distribution has sd == 0 and ",
+          "sd_sd == 0 then mean_sd must be 0, too."
+        )
+      }
+    } else {
+      if (missing(max)) {
+        stop("Maximum of parametric distributions must be specified.")
+      }
+    }
+  } else {
+    if (!all(missing(sd), missing(mean_sd),
+      missing(sd_sd), missing(dist), missing(max))) {
+        stop("If any distributional parameters are given then so must the mean.")
+    }
+  }
+
+  dist <- match.arg(dist)
+  if (fixed) {
     ret <- list(
       mean_mean = numeric(0),
       mean_sd = numeric(0),
       sd_mean = numeric(0),
       sd_sd = numeric(0),
-      fixed = integer(0),
-      dist = integer(0)
+      dist = integer(0),
+      max = integer(0)
     )
-    if (is.null(max)) {
-      ret$max <- integer(0)
-    } else {
-      ret$max <- max
+    if (length(pmf) == 0) {
+      if (missing(mean)) { ## empty
+        ret <- c(ret, list(
+          n = 0,
+          n_p = 0,
+          n_np = 0,
+          np_pmf_max = 0,
+          np_pmf = numeric(0),
+          np_pmf_groups = integer(0)
+        ))
+      } else { ## parametric fixed
+        if (sd == 0) { ## delta
+          pmf <- c(rep(0, mean), 1)
+        } else {
+          if (dist == "lognormal") {
+            params <- lognorm_dist_def(mean = mean, mean_sd = mean_sd,
+              sd = sd, sd_sd = sd_sd, max_value = max, samples = 1)
+          } else if (dist == "gamma") {
+            params <- gamma_dist_def(mean = mean, mean_sd = mean_sd,
+              sd = sd, sd_sd = sd_sd, max_value = max, samples = 1)
+          }
+          pmf <- dist_skel(
+            n = seq_len(max) - 1, dist = TRUE, cum = FALSE, model = dist,
+            params = params$params[[1]], max_value = max,
+            discrete = TRUE
+          )
+        }
+      }
+    } else { ## nonparametric fixed
+      if (!missing(max) && max > length(pmf)) {
+        pmf <- pmf[1:(max + 1)]
```

Comment:
This looks like it will cause an error from sight read.

Reply:
Fixed in https://github.com/epiforecasts/EpiNow2/pull/363/commits/25232c7da75861f497dda68dbb2e580be2da3cf4